runtime/common/src/xcm_impl.rs

```diff
@@ -0,0 +1,491 @@
+	match s {
+		CurrencyId::<R>::SelfReserve(_) => CurrencyId4Compare::SelfReserve,
+		CurrencyId::<R>::ParachainReserve(multi) =>
+			CurrencyId4Compare::ParachainReserve(multi.clone()),
+	}
+}
+
+impl<R: BaseRuntimeRequirements> Ord for CurrencyId<R> {
+	fn cmp(&self, other: &Self) -> Ordering {
+		convert_currency(self).cmp(&convert_currency(other))
+	}
+}
+
+impl<R: BaseRuntimeRequirements> PartialOrd for CurrencyId<R> {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl<R: BaseRuntimeRequirements> Default for CurrencyId<R> {
+	fn default() -> Self {
+		CurrencyId::ParachainReserve(Box::new(MultiLocation::here()))
+	}
+}
+
+/// Instructs how to convert a 32 byte accountId into a MultiLocation
+pub struct AccountIdToMultiLocation;
+impl spConvert<AccountId, MultiLocation> for AccountIdToMultiLocation {
+	fn convert(account: AccountId) -> MultiLocation {
+		MultiLocation {
+			parents: 0,
+			interior: X1(Junction::AccountId32 { network: NetworkId::Any, id: account.into() }),
+		}
+	}
+}
+
+pub struct OldAnchoringSelfReserve<R>(PhantomData<R>);
+impl<R: BaseRuntimeRequirements> OldAnchoringSelfReserve<R> {
+	/// Returns the value of this parameter type.
+	pub fn get() -> MultiLocation {
+		MultiLocation {
+			parents: 1,
+			interior: Junctions::X2(
+				Parachain(ParachainInfo::<R>::parachain_id().into()),
+				Junction::PalletInstance(<RuntimeBalances<R> as PalletInfoAccess>::index() as u8),
+			),
+		}
+	}
+}
+
+impl<I: From<MultiLocation>, R: BaseRuntimeRequirements> ::frame_support::traits::Get<I>
+	for OldAnchoringSelfReserve<R>
+{
+	fn get() -> I {
+		I::from(Self::get())
+	}
+}
+
+pub struct NewAnchoringSelfReserve<R>(PhantomData<R>);
+
+impl<R: BaseRuntimeRequirements> NewAnchoringSelfReserve<R> {
+	/// Returns the value of this parameter type.
+	pub fn get() -> MultiLocation {
+		MultiLocation {
+			parents: 0,
+			interior: Junctions::X1(Junction::PalletInstance(
+				<RuntimeBalances<R> as PalletInfoAccess>::index() as u8,
+			)),
+		}
+	}
+}
+
+impl<I: From<MultiLocation>, R: BaseRuntimeRequirements> ::frame_support::traits::Get<I>
+	for NewAnchoringSelfReserve<R>
+{
+	fn get() -> I {
+		I::from(Self::get())
+	}
+}
+
+impl<R: BaseRuntimeRequirements> From<MultiLocation> for CurrencyId<R> {
+	fn from(location: MultiLocation) -> Self {
+		match location {
+			a if (a == (OldAnchoringSelfReserve::<R>::get())) |
+				(a == (NewAnchoringSelfReserve::<R>::get())) =>
+				CurrencyId::<R>::SelfReserve(PhantomData::default()),
+			_ => CurrencyId::<R>::ParachainReserve(Box::new(location)),
+		}
+	}
+}
+
+impl<R: BaseRuntimeRequirements> From<Option<MultiLocation>> for CurrencyId<R> {
+	fn from(location: Option<MultiLocation>) -> Self {
+		match location {
+			Some(a)
+				if (a == OldAnchoringSelfReserve::<R>::get()) |
+					(a == NewAnchoringSelfReserve::<R>::get()) =>
+				CurrencyId::<R>::SelfReserve(PhantomData::default()),
+			Some(multi) => CurrencyId::ParachainReserve(Box::new(multi)),
```

Comment:
I believe here could be further simplified, like:
```suggestion
			Some(multi) => Self::from(multi),
```